File: HW/HW12/hw12.java
import java.util.Arrays;

public class hw12 {
    private int[] p;

    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        int[][] nums = Arrays.copyOf(pipes, pipes.length + n);
        for (int i = 0; i < n; i++) {
            nums[pipes.length + i] = new int[] {0, i + 1, wells[i]};
        }
        Arrays.sort(nums, (a, b) -> a[2] - b[2]);
        p = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            p[i] = i;
        }
        int ans = 0;
        for (var x : nums) {
            int a = x[0], b = x[1], c = x[2];
            int pa = find(a), pb = find(b);
            if (pa != pb) {
                p[pa] = pb;
                ans += c;
                if (--n == 0) {
                    return ans;
                }
            }
        }
        return ans;
    }

    private int find(int x) {
        if (p[x] != x) {
            p[x] = find(p[x]);
        }
        return p[x];
    }
    public static void main(String[] args) {
        hw12 solution = new hw12();
        int n = 2;
        int[] wells = {1, 1};
        int[][] pipes = {{1, 2, 1}, {1, 2, 2}};
        int result = solution.minCostToSupplyWater(n, wells, pipes);
        System.out.println("Minimum cost to supply water: " + result);
    }
}

File: HW/HW12/hw12.py
from typing import List

class hw12:
  def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
    #functions finds root of an element
    def find(x: int) -> int:
      if p[x] != x:
        p[x] = find(p[x])
      return p[x]

    for i, w in enumerate(wells, 1):
      #virtual edge from 0 node , with cost of well in house
      pipes.append([0, i, w])
    #sorts edges with costs
    pipes.sort(key=lambda x: x[2])
    #find sturcture initializeiation, each node is its own parent
    p = list(range(n + 1))
    ans = 0
    #go over each pipe iterating
    for a, b, c in pipes:
      pa, pb = find(a), find(b) #find the root of its parents
      if pa != pb:
        #if diff, add cost 
        p[pa] = pb
        #decrement # of nodes if connected
        n -= 1
        ans += c
        #if there are no more nodes left aka all connected 
        if n == 0:
          return ans
 
          
  def test_cases(self):
      solution = hw12()
      #given test case
      n = 2
      wells = [1, 1]
      pipes = [[1, 2, 1], [1, 2, 2]]
      print("Test Case Output:", solution.minCostToSupplyWater(n, wells, pipes))
      #expected output: 2

if __name__ == "__main__":
    hw12().test_cases()
